File: 3term/algos/lab4/src/bees_algo.rs
use crate::constants::{FORAGERS, SCOUTS};
use crate::graph::Graph;
use crate::section::Section;
use std::cmp::{max, min};
use std::collections::VecDeque;

pub struct BeesAlgorithm<'a> {
    pub graph: &'a Graph,
    pub sections: Vec<Section<'a>>,
    foragers: u32,
    scouts: u32,
}

impl<'a> BeesAlgorithm<'a> {
    pub fn new(g: &'a mut Graph) -> Self {
        Self {
            graph: g,
            sections: Vec::new(),
            foragers: FORAGERS,
            scouts: SCOUTS,
        }
    }

    fn dfs(&self, mut v: usize, mut color: u32) -> Section<'a> {
        let section = Section {
            graph: self.graph,
            coloring: vec![0; self.graph.get_no_vertices()],
            colors_used: 0,
        };

        let mut visited = vec![false; self.graph.get_no_vertices()];
        let mut queue = VecDeque::new();

        let mut last_section = section.clone();

        queue.push_back(section);

        while let Some(current_section) = queue.pop_front().as_mut() {
            visited[v] = true;
            current_section.coloring[v] = color;

            for (i, _) in current_section
                .graph
                .clone()
                .get_connected_vertices(v)
                .iter()
                .enumerate()
            {
                if visited[i] {
                    continue;
                }

                let mut j = 10;
                loop {
                    current_section.coloring[i] = j;

                    if current_section.is_valid_coloring() {
                        queue.push_back(current_section.clone());

                        v = i;
                        color = j;

                        break;
                    } else {
                        current_section.coloring[i] = 0;
                    }

                    j += 1;
                }
            }

            last_section = current_section.clone();
        }

        println!("{}", last_section.get_no_colors_used());

        last_section

        // for _, i := range section.graph.getConnectedVertexes(v) {
        // 	if !visited[i] {
        // 		for j := 10; ; j++ {
        // 			section.coloring[i] = j
        // 			if section.isValidColoring() {
        // 				dfs(i, j, visited, section)
        // 				break
        // 			} else {
        // 				section.coloring[i] = 0
        // 			}
        // 		}
        // 	}
        // }
    }

    pub fn gen_sections(&mut self) {
        for i in 0..min(
            max((SCOUTS * 4) as usize, self.sections.len()),
            self.graph.get_no_vertices(),
        ) {
            let mut p = self.dfs(i, 1);
            p.colors_used = p.get_no_colors_used();
            self.sections.push(p);
        }
    }
}

File: 3term/algos/lab4/bees_algo.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type BeeAlgorithm struct {
	graph    *Graph
	sections []*Section
	foragers int
	scouts   int
}

func (b *BeeAlgorithm) Start() {
	b.generateSections()

	for iteration := 0; iteration < ITERS_MAX; iteration++ {

		if iteration%20 == 0 {
			min := math.MaxInt32

			for _, section := range b.sections {
				if section != nil {
					min = Min(min, section.colorsUsed)
				}
			}

			fmt.Printf("%d: %d\n", iteration, min)
		}

		randSections := b.getRandomSections()

		totalColorsUsed := 0.0
		for _, s := range randSections {
			totalColorsUsed += float64(s.colorsUsed)
		}

		for _, s := range randSections {
			foragers := int(float64(b.foragers) * (1 - float64(s.colorsUsed)/totalColorsUsed))
			usedVertices := make([]bool, b.graph.getNOVertices())

			for foragers > 0 {
				unused := b.graph.getUnused(usedVertices)

				if len(unused) == 0 {
					break
				}

				maxVertex := unused[rand.Intn(len(unused))]
				usedVertices[maxVertex] = true

				for _, connectedVertex := range b.graph.getConnectedVertices(maxVertex) {
					if foragers <= 0 {
						break
					}

					s.coloring[maxVertex], s.coloring[connectedVertex] = s.coloring[connectedVertex], s.coloring[maxVertex]

					if s.isValidColoring() {
						foragers--
						currentColor := s.coloring[connectedVertex]

						for i := 1; i < currentColor; i++ {
							s.coloring[connectedVertex] = i

							if !s.isValidColoring() {
								s.coloring[connectedVertex] = currentColor
							} else {
								break
							}
						}
					} else {
						s.coloring[maxVertex], s.coloring[connectedVertex] = s.coloring[connectedVertex], s.coloring[maxVertex]
					}
				}
			}

			s.colorsUsed = s.getNOColorsUsed()
		}
	}

	min := math.MaxInt32
	for _, section := range b.sections {
		if section != nil {
			min = Min(min, section.colorsUsed)
		}
	}

	fmt.Printf("%d: %d\n", ITERS_MAX, min)
}

func (b *BeeAlgorithm) getRandomSections() []*Section {
	var sections []*Section
	selected := make(map[int]bool)

	for i := 0; i < Min(b.scouts, b.graph.getNOVertices()); i++ {
		for {
			index := rand.Intn(len(b.sections))

			if b.sections[index] == nil {
				continue
			}

			if _, ok := selected[index]; !ok {
				selected[index] = true
				sections = append(sections, b.sections[index])

				break
			}
		}
	}

	return sections
}

func dfs(v, color int, visited []bool, section *Section) {
	visited[v] = true
	section.coloring[v] = color

	for _, i := range section.graph.getConnectedVertices(v) {
		if visited[i] {
			continue
		}

		for j := 10; ; j++ {
			section.coloring[i] = j

			if section.isValidColoring() {
				dfs(i, j, visited, section)

				break
			} else {
				section.coloring[i] = 0
			}
		}
	}
}

func (b *BeeAlgorithm) generateSections() {
	for i := 0; i < Min(len(b.sections), b.graph.getNOVertices()); i++ {
		visited := make([]bool, b.graph.getNOVertices())
		section := &Section{
			graph:      b.graph,
			coloring:   make([]int, b.graph.getNOVertices()),
			colorsUsed: 0,
		}

		dfs(i, 1, visited, section)

		section.colorsUsed = section.getNOColorsUsed()
		b.sections[i] = section
	}
}

func NewBeeAlgorithm() *BeeAlgorithm {
	rand.Seed(time.Now().Unix())
	graph := genGraph(VERTICES)

	return &BeeAlgorithm{
		graph:    graph,
		sections: make([]*Section, SCOUTS*4),
		foragers: FORAGERS,
		scouts:   SCOUTS,
	}
}
